File: src/inference/sigma_attention.c
/* SPDX-License-Identifier: LicenseRef-SCSL-1.0 OR AGPL-3.0-only */
/*
 * v157 σ-attention inference (reference C).
 * Sub-quadratic pattern: full attention only for low-σ query rows; high-σ rows use a window.
 */
#include "sigma_attention.h"

#include <string.h>

static int32_t dot_prod(const int32_t *a, const int32_t *b, int32_t d)
{
    int64_t s;
    int32_t k;

    s = 0;
    for (k = 0; k < d; k++)
        s += (int64_t)a[k] * (int64_t)b[k];
    if (s > 2147483647LL)
        return 2147483647;
    if (s < -2147483647LL - 1LL)
        return -2147483647 - 1;
    return (int32_t)s;
}

static void swap_kv_row(int32_t *kv_cache, int32_t *sigmas, int32_t head_dim, int32_t a,
                        int32_t b)
{
    int32_t      stride;
    int32_t      i;
    int32_t      t;
    int32_t     *pa;
    int32_t     *pb;

    stride = head_dim * 2;
    pa     = kv_cache + (int32_t)(a * stride);
    pb     = kv_cache + (int32_t)(b * stride);
    for (i = 0; i < stride; i++) {
        t      = pa[i];
        pa[i]  = pb[i];
        pb[i]  = t;
    }
    t           = sigmas[a];
    sigmas[a]   = sigmas[b];
    sigmas[b]   = t;
}

static void one_row_attend(const int32_t *q_row, const int32_t *K, const int32_t *V,
                           int32_t d, int32_t j_start, int32_t j_end, int32_t *out_row)
{
    int64_t acc[COS_INFER_ATTN_DIM_MAX];
    int64_t sumw;
    int32_t j;
    int32_t k;

    if (d > COS_INFER_ATTN_DIM_MAX)
        return;

    for (k = 0; k < d; k++)
        acc[k]  = 0;
    sumw = 0;

    for (j = j_start; j < j_end; j++) {
        const int32_t *kj;
        const int32_t *vj;
        int32_t        s;
        int64_t        ws;

        kj = K + j * d;
        vj = V + j * d;
        s  = dot_prod(q_row, kj, d);
        if (s < 0)
            s = 0;
        ws = (int64_t)s + 1LL;
        sumw += ws;
        for (k = 0; k < d; k++)
            acc[k] += ws * (int64_t)vj[k];
    }

    if (sumw <= 0LL) {
        for (k = 0; k < d; k++)
            out_row[k] = 0;
        return;
    }

    for (k = 0; k < d; k++) {
        int64_t q;
        q = acc[k] / sumw;
        if (q > 2147483647LL)
            q = 2147483647LL;
        if (q < -2147483647LL - 1LL)
            q = -2147483647LL - 1LL;
        out_row[k] = (int32_t)q;
    }
}

void cos_infer_sigma_attention_init(cos_infer_sigma_attention_t *sa, int32_t seq_len,
                                    int32_t head_dim, int32_t n_heads, int32_t *token_sigmas,
                                    int32_t sigma_threshold_q16)
{
    if (!sa)
        return;
    memset(sa, 0, sizeof(*sa));
    sa->seq_len             = seq_len;
    sa->head_dim            = head_dim;
    sa->n_heads             = n_heads > 0 ? n_heads : 1;
    sa->token_sigmas        = token_sigmas;
    sa->sigma_threshold_q16 = sigma_threshold_q16;
}

void cos_infer_sigma_attention_reset_stats(cos_infer_sigma_attention_t *sa)
{
    if (!sa)
        return;
    sa->tokens_full   = 0U;
    sa->tokens_pruned = 0U;
}

void cos_infer_sigma_sparse_attention(cos_infer_sigma_attention_t *sa, const int32_t *Q,
                                      const int32_t *K, const int32_t *V, int32_t *output,
                                      int32_t window_size)
{
    int32_t n;
    int32_t d;
    int32_t i;
    int32_t w;

    if (!sa || !Q || !K || !V || !output || !sa->token_sigmas)
        return;
    n = sa->seq_len;
    d = sa->head_dim;
    if (d > COS_INFER_ATTN_DIM_MAX || n <= 0 || d <= 0)
        return;
    w = window_size > 0 ? window_size : 1;

    sa->tokens_full   = 0U;
    sa->tokens_pruned = 0U;

    for (i = 0; i < n; i++) {
        const int32_t *qi;
        int32_t       *oi;

        qi = Q + i * d;
        oi = output + i * d;

        if (sa->token_sigmas[i] < sa->sigma_threshold_q16) {
            one_row_attend(qi, K, V, d, 0, n, oi);
            sa->tokens_full++;
        } else {
            int32_t lo;
            int32_t hi;

            lo = i - w;
            hi = i + w + 1;
            if (lo < 0)
                lo = 0;
            if (hi > n)
                hi = n;
            one_row_attend(qi, K, V, d, lo, hi, oi);
            sa->tokens_pruned++;
        }
    }
}

int32_t cos_infer_sigma_kv_prune(cos_infer_sigma_attention_t *sa, int32_t *kv_cache,
                                 int32_t *sigmas, int32_t cache_len, int32_t head_dim,
                                 int32_t max_cache)
{
    int32_t worst_idx;
    int32_t worst_sig;
    int32_t i;
    int32_t new_len;

    if (!sa || !kv_cache || !sigmas || head_dim <= 0)
        return cache_len;
    if (cache_len <= max_cache)
        return cache_len;

    worst_idx = 0;
    worst_sig = sigmas[0];
    for (i = 1; i < cache_len; i++) {
        if (sigmas[i] > worst_sig) {
            worst_sig = sigmas[i];
            worst_idx = i;
        }
    }

    new_len = cache_len - 1;
    if (worst_idx != new_len)
        swap_kv_row(kv_cache, sigmas, head_dim, worst_idx, new_len);
    sa->tokens_pruned++;
    return new_len;
}

void cos_infer_sigma_attention_ops_estimate(const cos_infer_sigma_attention_t *sa,
                                            int32_t seq_len, int32_t window_size, int32_t head_dim,
                                            int64_t *out_sparse_ops, int64_t *out_dense_ops)
{
    int64_t dense;
    int64_t sparse;
    uint32_t f;
    uint32_t p;
    int32_t w;
    int32_t span;

    if (!out_sparse_ops || !out_dense_ops)
        return;

    dense = (int64_t)seq_len * (int64_t)seq_len * (int64_t)head_dim;
    *out_dense_ops = dense;

    if (!sa) {
        *out_sparse_ops = dense;
        return;
    }

    f = sa->tokens_full;
    p = sa->tokens_pruned;
    w = window_size > 0 ? window_size : 1;
    span = w * 2 + 1;
    if (span > seq_len)
        span = seq_len;

    sparse = (int64_t)f * (int64_t)seq_len * (int64_t)head_dim
             + (int64_t)p * (int64_t)span * (int64_t)head_dim;
    *out_sparse_ops = sparse;
}

File: python/cos/sigma_attention.py
# SPDX-License-Identifier: LicenseRef-SCSL-1.0 OR AGPL-3.0-only
# Copyright (c) 2024-2026 Lauri Elias Rainio and Spektre Labs Oy.
# All rights reserved. See LICENSE for binding terms.
"""
σ-attention (lab): sub-quadratic attention by applying full pairwise weights only on **low-σ**
query rows; high-σ rows use a sliding window.

Python implementation mirrors ``src/inference/sigma_attention.c`` semantics for CI / CLI; it is not
FlashAttention and does not fuse GPU kernels.
"""
from __future__ import annotations

import math
import time
from typing import Any, Dict, List, Optional, Protocol, Sequence, Tuple

COS_ATTN_Q16 = 65536


class TokenSigmaGate(Protocol):
    def estimate_token_sigma(self, index: int, *, seq_len: int) -> float:
        """Per-position σ in [0, 1] — lower means “trust this token more” (full attention)."""


class LabTokenSigmaGate:
    """Deterministic demo: ~``low_fraction`` of indices get low σ (full attention candidates)."""

    def __init__(self, seq_len: int, *, low_fraction: float = 0.12, low_sigma: float = 0.08, high_sigma: float = 0.72):
        self.seq_len = int(seq_len)
        self.low_fraction = float(low_fraction)
        self.low_sigma = float(low_sigma)
        self.high_sigma = float(high_sigma)

    def estimate_token_sigma(self, index: int, *, seq_len: int) -> float:
        n = max(1, int(seq_len))
        cutoff = max(1, int(math.floor(self.low_fraction * n)))
        if int(index) < cutoff:
            return self.low_sigma
        return self.high_sigma


def _dot_rows(a: Sequence[float], b: Sequence[float]) -> float:
    return sum(float(x) * float(y) for x, y in zip(a, b))


def _attend_row(
    q_row: Sequence[float],
    k_rows: Sequence[Sequence[float]],
    v_rows: Sequence[Sequence[float]],
    j_start: int,
    j_end: int,
) -> List[float]:
    d = len(q_row)
    acc = [0.0] * d
    sumw = 0.0
    for j in range(max(0, j_start), min(len(k_rows), j_end)):
        s = _dot_rows(q_row, k_rows[j])
        w = max(0.0, s) + 1.0
        sumw += w
        for k in range(d):
            acc[k] += w * float(v_rows[j][k])
    if sumw <= 0.0:
        return [0.0] * d
    return [acc[k] / sumw for k in range(d)]


class SigmaAttention:
    def __init__(
        self,
        gate: Optional[Any] = None,
        *,
        window_size: int = 128,
        sigma_threshold: float = 0.3,
    ) -> None:
        self.gate = gate
        self.window = max(1, int(window_size))
        self.threshold = float(sigma_threshold)
        self.stats: Dict[str, int] = {"full": 0, "pruned": 0}

    def forward(
        self,
        q: Sequence[Sequence[float]],
        k: Sequence[Sequence[float]],
        v: Sequence[Sequence[float]],
        *,
        token_sigmas: Optional[Sequence[float]] = None,
    ) -> List[List[float]]:
        seq_len = len(q)
        g = self.gate
        self.stats = {"full": 0, "pruned": 0}
        sigs: List[float]
        if token_sigmas is not None:
            sigs = [float(x) for x in token_sigmas[:seq_len]]
            while len(sigs) < seq_len:
                sigs.append(0.5)
        else:
            if g is None:
                g = LabTokenSigmaGate(seq_len)
            sigs = [float(g.estimate_token_sigma(i, seq_len=seq_len)) for i in range(seq_len)]

        thr = self.threshold
        out: List[List[float]] = []
        for i in range(seq_len):
            qs = float(sigs[i])
            if qs < thr:
                out.append(_attend_row(q[i], k, v, 0, seq_len))
                self.stats["full"] += 1
            else:
                lo = max(0, i - self.window)
                hi = min(seq_len, i + self.window + 1)
                out.append(_attend_row(q[i], k, v, lo, hi))
                self.stats["pruned"] += 1
        return out

    def complexity(self, seq_len: int) -> Dict[str, Any]:
        total = self.stats["full"] + self.stats["pruned"]
        if total <= 0:
            return {"estimated_savings": "0%", "full_ratio": 0.0, "full_tokens": 0, "pruned_tokens": 0}
        fr = self.stats["full"] / float(total)
        eff = fr * float(seq_len) + (1.0 - fr) * float(2 * self.window + 1)
        baseline = float(seq_len)
        saving = 0.0
        if baseline > 1e-9:
            saving = max(0.0, 100.0 * (1.0 - eff / baseline))
        return {
            "effective_key_span_per_row": round(eff, 4),
            "full_ratio": round(fr, 6),
            "full_tokens": int(self.stats["full"]),
            "pruned_tokens": int(self.stats["pruned"]),
            "estimated_savings_vs_dense_n": f"{saving:.0f}% sub-quadratic pattern",
        }


def dense_attention_reference(q: Sequence[Sequence[float]], k: Sequence[Sequence[float]], v: Sequence[Sequence[float]]) -> List[List[float]]:
    """O(N²) reference (same weighting rule as :meth:`SigmaAttention.forward` full rows)."""
    n = len(q)
    return [_attend_row(q[i], k, v, 0, n) for i in range(n)]


def benchmark_sigma_vs_dense(
    seq_len: int,
    head_dim: int,
    *,
    window: int = 128,
    threshold: float = 0.3,
    low_fraction: float = 0.12,
    repeats: int = 2,
) -> Dict[str, Any]:
    """Wall-clock lab ratio; not hardware FlashAttention."""
    n = max(2, int(seq_len))
    d = max(1, int(head_dim))
    q = [[float((i + k) % 17) / 16.0 for k in range(d)] for i in range(n)]
    kk = [[float((i * k + 3) % 19) / 18.0 for k in range(d)] for i in range(n)]
    vv = [[float((i + k * k) % 13) / 12.0 for k in range(d)] for i in range(n)]

    gate = LabTokenSigmaGate(n, low_fraction=low_fraction)
    attn = SigmaAttention(gate, window_size=window, sigma_threshold=threshold)

    t0 = time.perf_counter()
    for _ in range(repeats):
        dense_attention_reference(q, kk, vv)
    t_dense = max(1e-12, (time.perf_counter() - t0) / float(max(1, repeats)))

    t1 = time.perf_counter()
    for _ in range(repeats):
        attn.forward(q, kk, vv)
    t_sparse = max(1e-12, (time.perf_counter() - t1) / float(max(1, repeats)))

    pruned_ratio = attn.stats["pruned"] / float(n)
    return {
        "dense_ms_per_run": round(t_dense * 1000.0, 4),
        "head_dim": d,
        "pruned_tokens": int(attn.stats["pruned"]),
        "seq_len": n,
        "sparse_ms_per_run": round(t_sparse * 1000.0, 4),
        "speedup": round(t_dense / t_sparse, 4),
        "full_tokens": int(attn.stats["full"]),
        "pruned_token_ratio": round(pruned_ratio, 4),
    }


def kv_prune_rows(
    kv_rows: List[List[float]],
    sigmas: List[float],
    *,
    max_cache: int,
) -> Tuple[Tuple[List[List[float]], List[float]], int]:
    """
    Drop one highest-σ cache row (swap with last), mirroring ``cos_infer_sigma_kv_prune``.
    Each ``kv_rows[i]`` is one flattened row ``K_i || V_i`` (2 * head_dim floats).
    Returns ``(new_kv, new_sigmas), pruned_count`` (single step).
    """
    if len(kv_rows) != len(sigmas) or len(kv_rows) <= max_cache:
        return (kv_rows, sigmas), 0
    worst = 0
    worst_s = float(sigmas[0])
    for i in range(1, len(sigmas)):
        if float(sigmas[i]) > worst_s:
            worst_s = float(sigmas[i])
            worst = i
    last = len(kv_rows) - 1
    kv = [list(r) for r in kv_rows]
    sg = list(sigmas)
    if worst != last:
        kv[worst], kv[last] = kv[last], kv[worst]
        sg[worst], sg[last] = sg[last], sg[worst]
    return (kv[:-1], sg[:-1]), 1


__all__ = [
    "COS_ATTN_Q16",
    "LabTokenSigmaGate",
    "SigmaAttention",
    "benchmark_sigma_vs_dense",
    "dense_attention_reference",
    "kv_prune_rows",
]
